src/main/java/com/iota/iri/service/milestone/impl/MilestoneSolidifierImpl.java

```diff
@@ -1,339 +1,666 @@
- * issues checkSolidity calls on the earliest milestones to solidify them.
- * </p>
- * <p>
- * To save resources and make the call a little bit more efficient, we cache the earliest milestones in a separate map,
- * so the relatively expensive task of having to search for the next earliest milestone in the pool only has to be
- * performed after a milestone has become solid or irrelevant for our node.
- * </p>
- */
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ArrayBlockingQueue;
+
 public class MilestoneSolidifierImpl implements MilestoneSolidifier {
-    /**
-     * Defines the amount of milestones that we "simultaneously" try to solidify in one pass.
-     */
-    private static final int SOLIDIFICATION_QUEUE_SIZE = 20;
+    private static final IntervalLogger log = new IntervalLogger(MilestoneSolidifierImpl.class);
+    // Max size fo the solidification queue
+    private static final int MAX_SIZE = 10;
 
-    /**
-     * Defines the interval in which solidity checks are issued (in milliseconds).
-     */
-    private static final int SOLIDIFICATION_INTERVAL = 100;
+    private Map<Hash, Integer> unsolidMilestones = new ConcurrentHashMap<>();
+    private Map<Hash, Integer> solidificationQueue = new ConcurrentHashMap<>();
+    private Map<Integer, Hash> seenMilestones = new ConcurrentHashMap<>();
+
+    private Map.Entry<Hash, Integer> oldestMilestoneInQueue = null;
+
+    private int latestMilestoneIndex;
+    private Hash latestMilestoneHash;
+    private int latestSolidMilestone;
+
+    private TransactionSolidifier transactionSolidifier;
+    private LedgerService ledgerService;
+    private SnapshotProvider snapshotProvider;
+    private Tangle tangle;
+    private TransactionRequester transactionRequester;
+    private MilestoneService milestoneService;
+    private SolidificationConfig config;
+
+
+    private SyncProgressInfo syncProgressInfo = new SyncProgressInfo();
 
     /**
-     * Logger for this class allowing us to dump debug and status messages.
+     * Holds the milestone index of the milestone that caused the repair logic to get started.
      */
-    private static final IntervalLogger log = new IntervalLogger(MilestoneSolidifier.class);
+    private int errorCausingMilestoneIndex = Integer.MAX_VALUE;
 
     /**
-     * Holds the snapshot provider which gives us access to the relevant snapshots.
+     * Counter for the backoff repair strategy (see {@link #repairCorruptedMilestone(MilestoneViewModel)}.
      */
-    private final SnapshotProvider snapshotProvider;
+    private int repairBackoffCounter = 0;
 
     /**
-     * Holds a reference to the transactionSolidifier which allows us to issue solidity checks.
+     * An indicator for whether or not the solidifier has processed from the first batch of seen milestones
      */
-    private final TransactionSolidifier transactionSolidifier;
+    private boolean initialized = false;
```

Comment:
also a bit scary

Reply:
Also switched to atomic, this was there before as well.